kernel/tests/integration-test.rs

```diff
@@ -1754,11 +1803,13 @@ fn s04_userspace_rumprt_net() {
 
     let mut output = String::new();
     let mut qemu_run = || -> Result<WaitStatus> {
+        // need to spawn nrk first to set up tap interface
+        let mut p = spawn_nrk(&cmdline)?;
+        std::thread::sleep(std::time::Duration::from_secs(2));
```

Comment:
Is the sleep because now you're removing adding interfaces and dhcpd complains if it starts up too early before the interface isn't there?

Reply:
Yes, that's correct. I probably should have added a comment

Reply:
Can we
- either move the sleep it into `spawn_nrk` (that will slightly slow down tests that don't use dhcp, but will not cause unexpected issues with future writers of tests that do use dhcp and forget to add the sleep)
- Or maybe have the network setup happen with a separate invocation to run.py (I *think* tests already call run.py twice once with --no-run and once without) .. If we go down this route, I fear this will complicate running tests manually a bit which is unfortunate, maybe the regular run.py should still check the setup and abort with an error if it detects that the tap interfaces aren't configured like they should be for this invocation then...)?

what do you think?

Reply:
I think adding the pause to spawn_nrk() would be sufficient... we could also make it a configurable parameter to spawn_nrk for whether it should wait or not if we went that route. I think this option is the simplest.

If the test already calls run.py with no-run we could pretty easily add a "setup-network" flag to indicate whether run.py should configure the network or not (maybe another parameter of the net subparser?). Seems a shame to run network teardown/setup twice if the tests are indeed calling run.py twice. So we could create the network setup on only one of the calls, maybe? e.g., the first (or an additional) call to run.py?

A nice thing about option 2 is that, unllike option 1, it would allow you to start dhcp before nrk, which is what the test originally did.

I also want to ask - does the network setup information belong in run.py? I agree some basic verification (do taps/bridges exist) seems wise, but if we're separating network functionality from the running of nrk (by setting up in a seperate call to run.py), maybe network config belongs in some other script (setup_network.py or something). Seems a bit odd to have to use a complicated script like run.py with lots of args for a task that will ignore most of those args.

I think I like option 1 (for simplicity) or option 3 (separate script) best. But I'm open to any of the three options

Reply:
Agree, run.py is definitely becoming a bit too complicated these days :)

I think -- for now -- what you suggested just moving the sleep in spawn_nrk is probably fine (with some comment).. IMO No need to make it optional -- we waste a few seconds but that's ok the test run for 40mins anyways.